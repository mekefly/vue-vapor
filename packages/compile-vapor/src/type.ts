export type CodeSnippet = {
  id: number;
  createSnippet: string;
  attributeSnippet: string;
  mount: string;
  parentId: number;
};
