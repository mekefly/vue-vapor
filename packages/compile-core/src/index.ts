export * from "./parser";
export * from "./traverser";
export function xxxx(params: string) {
  console.log("hello");
}
xxxx("33");
