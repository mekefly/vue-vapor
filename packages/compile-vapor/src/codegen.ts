import { SFC } from "@vue-vapor/compile-sfc";
import { createSnippetOptionsByAst } from "./snippet/createSnippet";
import {
  moduleTemplate,
  renderCodeSnippetTemplate,
  setupSnippetTemplate,
} from "./template";
import { MarginSnippetOptions } from "./type";

export function codegen(sfc: SFC) {
  if (!(sfc.setup && sfc.vapor))
    throw new Error("这里无法处理，你可以使用 `@vue-vapor/compile-core` 试试");

  const snippetsOptions = createSnippetOptionsByAst(sfc.template);
  const importSnippet = sfc.importSnippets.join("\n");
  const scriptSnippet = sfc.script;
  const code = marginSnippet({
    importSnippet,
    scriptSnippet,
    ...snippetsOptions,
  });
  return format(code);
}

export function format(code: string) {
  return code;
}

export function marginSnippet(options: MarginSnippetOptions) {
  const {
    createSnippet,
    appendSnippet,
    attributeSnippet,
    importSnippet,
    scriptSnippet,
  } = options;
  return moduleTemplate(
    importSnippet,
    setupSnippetTemplate(
      scriptSnippet,
      renderCodeSnippetTemplate(createSnippet, appendSnippet, attributeSnippet)
    )
  );
}
