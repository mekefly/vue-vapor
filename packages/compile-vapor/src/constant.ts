export const CONTEXT_VAR_NAME = "__context";

export const CREATE_ELEMENT_FUNCTION_NAME = "ce";
export const CREATE_TEXT_FUNCTION_NAME = "ct";
export const CREATE_COMPONENT_FUNCTION_NAME = "cc";

export const ADD_EVENT_LISTENER_FUNCTION_NAME = "ae";
export const SET_ATTRIBUTE_FUNCTION_NAME = "sa";

export const RENDER_VAR_NAME = `$render`;

export const ELEMENT_ARRAY_VAR_NAME = "$";
export const EFFECT_FUNCTION_NAME = "effect";
