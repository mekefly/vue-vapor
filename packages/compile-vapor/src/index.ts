import {
  AstNode,
  HTMLElementAst,
  HTMLTextAst,
  NodeType,
  typeFor,
} from "@vue-vapor/compile-core";
import { SFC } from "@vue-vapor/compile-sfc";
import {
  appendTemplate,
  createElementTemplate,
  createSnippetTemplate,
  createTextTemplate,
  createWatchEffectSnippet,
  moduleSnippetTemplate,
  renderCodeSnippetTemplate,
  setAttributeTemplate,
  setupSnippetTemplate,
} from "./template";
import { CodeSnippet } from "./type";

export function codegen(sfc: SFC) {
  if (!(sfc.setup && sfc.vapor))
    throw new Error("这里无法处理，你可以使用 `@vue-vapor/compile-core` 试试");

  const snippetsOptions = createSnippetsByAst(sfc.template);
  return marginSFC(sfc, snippetsOptions as any);
}

function marginSFC(
  sfc: SFC,
  { createSnippet, appendSnippet, attributeSnippet }: RenderSnippetOption
) {
  return moduleSnippetTemplate(
    sfc.importSnippets.join("\n"),
    setupSnippetTemplate(
      sfc.script,
      renderCodeSnippetTemplate(createSnippet, appendSnippet, attributeSnippet)
    )
  );
}

const parentIdMap = new Map<any, number>();

type RenderSnippetOption = {
  createSnippet: string;
  appendSnippet: string;
  attributeSnippet: string;
};
function createSnippetsByAst(ast: HTMLElementAst): RenderSnippetOption {
  const astList = genAstList(ast);

  const astNodeList = genAstNodeList(astList);

  return createSnippets(astNodeList);
}

function genAstNodeList(astList: AstNode[]) {
  const { HTMLElementAst, HTMLTemplateStatementAst, HTMLTextAst } = NodeType;
  return astList.map((ast, id) => {
    if (typeFor(ast, HTMLElementAst)) {
      return createCodeSnippetByEl(ast, id);
    } else if (typeFor(ast, HTMLTextAst)) {
      return createCodeSnippetByText(ast, id);
    } else if (typeFor(ast, HTMLTemplateStatementAst)) {
      throw new Error("todo");
    }
    throw new Error("不适合的标签");
  });
}

function createSnippets(codeSnippets: CodeSnippet[]) {
  // 创建语句
  const createSnippet = createSnippetTemplate(codeSnippets);

  // append生成
  const appendSnippet = genAppendSnippet(codeSnippets);

  // 属性设置
  const attributeSnippet = codeSnippets
    .map((codeSnippet) => codeSnippet.attributeSnippet)
    .join("");
  return { createSnippet, appendSnippet, attributeSnippet };
}

function genAppendSnippet(codeSnippets: CodeSnippet[]) {
  const appendMap: number[][] = [];
  codeSnippets.forEach((codeSnippet) => {
    const appends = (appendMap[codeSnippet.parentId] =
      appendMap[codeSnippet.parentId] ?? []);
    appends.push(codeSnippet.id);
  });

  const appendSnippet = appendMap
    .map((ids, parentId) => appendTemplate(parentId, ...ids))
    .join("\n");
  return appendSnippet;
}

function genAstList(ast: HTMLElementAst) {
  const bfsAstList: AstNode[] = [...ast.children];
  bfsAstList.forEach((ast, i) => {
    const id = i + 1;
    const children = ast.getChildren();
    if (!children) {
      return;
    }
    children.forEach((child) => {
      bfsAstList.push(child);
      parentIdMap.set(child, id);
    });
  });
  return bfsAstList;
}

function createCodeSnippetByEl(ast: HTMLElementAst, id: number): CodeSnippet {
  const parentId = parentIdMap.get(ast) ?? 0;

  return {
    parentId,
    id,
    createSnippet: createElementTemplate(ast, id),
    attributeSnippet: createWatchEffectSnippet(
      Object.entries(ast.props)
        .map(([key, value]) => setAttributeTemplate(id, key, value))
        .join("")
    ),
    mount: appendTemplate(parentId, id),
  };
}
function createCodeSnippetByText(ast: HTMLTextAst, id: number) {
  const parentId = parentIdMap.get(ast) ?? 0;
  return {
    parentId,
    id,
    createSnippet: createTextTemplate(ast, id),
    attributeSnippet: ``,
    mount: appendTemplate(parentId, id),
  };
}
