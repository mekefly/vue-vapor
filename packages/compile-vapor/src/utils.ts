export const INDENT_UNIT = 2;

/**
 * 控制代码缩进
 *
 * @export
 * @param {string} snippets 代码段
 * @param {number} num 正数为增加缩进，负数为减少缩进
 */
export function indent(snippets: string, num: number = 1) {
  const isPN = num >= 0;
  const spaces = space((num = isPN ? num : -num) * INDENT_UNIT);

  const callback: (s: string) => string = isPN
    ? // 正数增加空格
      (snippet) => `${spaces}${snippet}`
    : // 负数减少空格
      (snippet) => (snippet.startsWith(spaces) ? snippet.slice(-num) : snippet);

  return snippets.split("\n").map(callback).join("\n");
}
export function space(num: number) {
  return "".padStart(num, " ");
}

export function spaceDisplay(str: string, invalid: boolean = false) {
  if (invalid) return;
  return str.replaceAll(" ", "_");
}

export function ignoreEndWithEnter(text: string) {
  return text.endsWith("\n") ? text.slice(0, -1) : text;
}
export function ignoreStartWithEnter(text: string) {
  return text.startsWith("\n") ? text.slice(1) : text;
}
export function filterNoUndefined<A extends Array<any>>(
  arr: A
): A extends Array<infer P> ? (P extends undefined ? never : Array<P>) : never {
  return arr.filter((e) => e !== undefined) as any;
}

type xxx = "333" & "222";
