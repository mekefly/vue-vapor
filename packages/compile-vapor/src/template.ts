import { HTMLElementAst, HTMLTextAst } from "@vue-vapor/compile-core";
import { CodeSnippet } from "./type";
import { indent } from "./utils";

export const SET_ATTRIBUTE_VAR_NAME = "sa";
export const RENDER_VAR_NAME = `$render`;
export const VARIABLE_NAME = "node";
export const EFFECT_NAME = "effect";

const compileImport: Record<string, Set<string>> = {};

export function moduleSnippetTemplate(
  importSnippet: string,
  setupSnippet: string
) {
  importSnippet =
    Object.entries(compileImport)
      .map(
        ([packageName, vars]) =>
          `import { ${[...vars].join(",")} } from "@vue/reactivity";\n`
      )
      .join() + importSnippet;

  return `${importSnippet ? `${importSnippet}\n\n` : "\n"}${setupSnippet}`;
}

export function setupSnippetTemplate(
  scriptSnippet: string,
  renderSnippet: string
) {
  return `export default function (props,context){
  ${indent(scriptSnippet, 1)}

  ${indent(renderSnippet)}
}`;
}

export function renderCodeSnippetTemplate(
  createSnippet: string,
  appendSnippet: string,
  attributeSnippet: string
) {
  return `
  ${createSnippet}
  ${appendSnippet}
  ${createWatchEffectSnippet(attributeSnippet)}`;
}

export function createSnippetTemplate(codeSnippets: CodeSnippet[]) {
  return `var ${VARIABLE_NAME} = {${codeSnippets
    .map((codeSnippet) => {
      return codeSnippet.createSnippet;
    })
    .join(",")}};`;
}

export function appendTemplate(parentId: number, ...ids: number[]) {
  return `${getElVarTemplate(parentId)}.append(${ids
    .map((id) => getElVarTemplate(id))
    .join(",")});`;
}

const saTemplate = `function ${SET_ATTRIBUTE_VAR_NAME}(e, key, value) {e.setAttribute(key, value);}`;
export function setAttributeTemplate(id: number, key: string, value: string) {
  const dyn = ["@", ":", "v-"].some((item) => key.startsWith(item));
  addImport("@vue/reactivity", "unref");

  if (key.startsWith(":")) {
    return `${getElVarTemplate(id)}.setAttribute("${key.slice(
      1
    )}",${`unref(${value})`});`;
  } else if (key.startsWith("@")) {
    return `${getElVarTemplate(id)}.addEventListener("${key.slice(
      1
    )}",${value});`;
  } else {
    return `${getElVarTemplate(id)}.setAttribute("${key}","${value}"});`;
  }
}

export function getElVarTemplate(id: number) {
  return `${VARIABLE_NAME}.$${id}`;
}

export function createElementTemplate(ast: HTMLElementAst, id: number) {
  return `$${id}: document.createElement("${ast.tag}")`;
}

export function createTextTemplate(ast: HTMLTextAst, id: number) {
  return `$${id}: document.createTextNode(${JSON.stringify(ast.text)})`;
}

export function createWatchEffectSnippet(snippet: string) {
  if (!snippet) return "";
  addImport("@vue/reactivity", EFFECT_NAME);
  return `${EFFECT_NAME}(()=>{${snippet}});`;
}
export function addImport<T extends `@${`vue${`/reactivity` | `-vapor`}`}`>(
  packageName: T,
  ...varName: string[]
) {
  const compileImportDeps =
    compileImport[packageName] ?? (compileImport[packageName] = new Set());

  varName.forEach((item) => compileImportDeps.add(item));
}
