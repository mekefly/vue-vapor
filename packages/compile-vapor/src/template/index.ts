import { HTMLElementAst, HTMLTextAst } from "@vue-vapor/compile-core";
import {
  CONTEXT_VAR_NAME,
  CREATE_COMPONENT_FUNCTION_NAME,
  CREATE_ELEMENT_FUNCTION_NAME,
  CREATE_TEXT_FUNCTION_NAME,
  EFFECT_FUNCTION_NAME,
  ELEMENT_ARRAY_VAR_NAME,
} from "../constant";
import { AstCodeSnippetOptions, ModuleImport } from "../type";
import { ignoreEndWithEnter, indent } from "../utils";
import {
  addImport,
  createImportTemplate,
  moduleImport,
  moduleSnippets,
} from "./moduleTemplate";

export function moduleTemplate(importSnippet: string, setupSnippet: string) {
  importSnippet = marginImportSnippet(importSnippet, moduleImport);
  return `\
${ignoreEndWithEnter(importSnippet)}\n\n${[...moduleSnippets].join(
    "\n"
  )}${setupSnippet}\
`;
}

export function marginImportSnippet(
  importSnippet: string,
  moduleImport: ModuleImport
): string {
  return `\
${Object.entries(moduleImport)
  .map(([packageName, vars]) => createImportTemplate(vars, packageName))
  .join()}\
${importSnippet}`;
}

// prettier-ignore
export function setupSnippetTemplate(
  scriptSnippet: string,
  renderSnippet: string
) {
  return `\
export default function (props,${CONTEXT_VAR_NAME}){
${indent(`${ignoreEndWithEnter(scriptSnippet)}
(function(){
${indent(renderSnippet)}
})()`)}
}`;
}

// prettier-ignore
export function renderCodeSnippetTemplate(
  createSnippet: string,
  appendSnippet: string,
  attributeSnippet: string
) {
  return `\
${createSnippet}
${appendSnippet}
${ignoreEndWithEnter(attributeSnippet)}`;
}

export function createSnippetTemplate(codeSnippets: AstCodeSnippetOptions[]) {
  return `\
const ${ELEMENT_ARRAY_VAR_NAME} = ${CONTEXT_VAR_NAME}.el;
${codeSnippets.map((codeSnippet) => codeSnippet.createSnippet).join("")}\
`;
}

export function appendTemplate(parentId: number, ...ids: number[]) {
  return `${getElVarTemplate(parentId)}.append(${ids
    .map((id) => getElVarTemplate(id))
    .join(",")});\n`;
}

export function getElVarTemplate(id: number) {
  return `${ELEMENT_ARRAY_VAR_NAME}[${id}]`;
}


export function createComponentTemplate(
  ComponentName: string,
  parentId: number,
  id: number
) {
  return setElTemplate(
    id,
    `${CREATE_COMPONENT_FUNCTION_NAME}(${ComponentName},${getElVarTemplate(
      parentId
    )})`
  );
}

export function createTextTemplate(ast: HTMLTextAst, id: number) {
  return setElTemplate(
    id,
    `${CREATE_TEXT_FUNCTION_NAME}(${JSON.stringify(ast.text)})`
  );
}

export function createElementTemplate(ast: HTMLElementAst, id: number) {
  return setElTemplate(id, `${CREATE_ELEMENT_FUNCTION_NAME}("${ast.tag}")`);
}

export function createTemplateStatementTemplate(
  id: number,
  defaul: string = '""'
) {
  return setElTemplate(id, `${CREATE_TEXT_FUNCTION_NAME}(${defaul})`);
}

export function setElTemplate(id: number, snip: string) {
  return `${ELEMENT_ARRAY_VAR_NAME}[${id}] = ${snip};\n`;
}
