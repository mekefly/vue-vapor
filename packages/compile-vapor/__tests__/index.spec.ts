import { readFileSync, writeFileSync } from "fs";
import { resolve } from "path";
import { parser } from "../../compile-core";
import { compileSFC } from "../../compile-sfc";
import { codegen, genAstList } from "../src";
describe("index", () => {
  function getSFC(path1: string) {
    const path = resolve(__dirname, `./case/`, `${path1}.vue`);
    const code = readFileSync(path, "utf-8");
    const ast = parser(code);
    return compileSFC(ast);
  }
  function writeCode(name: string, code: string) {
    const path = resolve(__dirname, `./case/`, `${name}.ts`);
    writeFileSync(path, code);
  }
  describe("genAstList", () => {
    test("<div></div>", ({ meta: { name } }) => {
      const ast = parser(name).getChildren()[0] as any;
      const list = genAstList(ast);
      expect(list).toMatchInlineSnapshot(`
      [
        HTMLElementAst {
          "children": [],
          "props": {},
          "tag": "div",
          "type": "HTMLElementAst",
        },
      ]
    `);
    });
    test("<div><div></div></div>", ({ meta: { name } }) => {
      const ast = parser(name).getChildren()[0] as any;
      const list = genAstList(ast);
      expect(list).toMatchInlineSnapshot(`
        [
          HTMLElementAst {
            "children": [
              HTMLElementAst {
                "children": [],
                "props": {},
                "tag": "div",
                "type": "HTMLElementAst",
              },
            ],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
          HTMLElementAst {
            "children": [],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ]
      `);
    });
  });

  test("<template>\n<div></div></template>", ({ meta: { name } }) => {
    const ast = parser(name).getChildren()[0] as any;
    const list = genAstList(ast);
    expect(list.length).toMatchInlineSnapshot("3");
    expect(list).toMatchInlineSnapshot(`
      [
        HTMLElementAst {
          "children": [
            HTMLTextAst {
              "text": "
      ",
              "type": "HTMLTextAst",
            },
            HTMLElementAst {
              "children": [],
              "props": {},
              "tag": "div",
              "type": "HTMLElementAst",
            },
          ],
          "props": {},
          "tag": "template",
          "type": "HTMLElementAst",
        },
        HTMLTextAst {
          "text": "
      ",
          "type": "HTMLTextAst",
        },
        HTMLElementAst {
          "children": [],
          "props": {},
          "tag": "div",
          "type": "HTMLElementAst",
        },
      ]
    `);
  });
  test("<template><div><div><div><div><div></div></div></div></div></div></template>", ({
    meta: { name },
  }) => {
    const ast = parser(name).getChildren()[0] as any;
    const list = genAstList(ast);
    expect(list.length).toMatchInlineSnapshot("6");
    expect(list).toMatchSnapshot();
  });
  test("<template><div><div><div><div><div></div></div></div></div></div></template>", ({
    meta: { name },
  }) => {
    const ast = parser(name).getChildren()[0] as any;
    const list = genAstList(ast);
    expect(list.length).toMatchInlineSnapshot("6");
    expect(list).toMatchSnapshot();
  });
  test("case2", ({ meta: { name } }) => {
    const sfc = getSFC(name);
    expect(sfc).toMatchInlineSnapshot(`
      {
        "importSnippets": [
          "import { ref } from \\"@vue/reactivity\\"",
        ],
        "script": "const count = ref(0);
      count.value++;
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      ",
        "setup": true,
        "style": "",
        "template": HTMLElementAst {
          "children": [
            HTMLTextAst {
              "text": "
        ",
              "type": "HTMLTextAst",
            },
            HTMLElementAst {
              "children": [
                HTMLTextAst {
                  "text": "
          ",
                  "type": "HTMLTextAst",
                },
                HTMLElementAst {
                  "children": [
                    HTMLElementAst {
                      "children": [],
                      "props": {
                        ":value": "count",
                      },
                      "tag": "input",
                      "type": "HTMLElementAst",
                    },
                  ],
                  "props": {},
                  "tag": "div",
                  "type": "HTMLElementAst",
                },
                HTMLTextAst {
                  "text": "
        ",
                  "type": "HTMLTextAst",
                },
              ],
              "props": {},
              "tag": "div",
              "type": "HTMLElementAst",
            },
            HTMLTextAst {
              "text": "
      ",
              "type": "HTMLTextAst",
            },
          ],
          "props": {},
          "tag": "template",
          "type": "HTMLElementAst",
        },
        "vapor": true,
      }
    `);

    const list = genAstList(sfc.template);
    expect(list.length).toMatchInlineSnapshot("8");
    expect(list.map((item) => item.type + " => " + Object.values(item)))
      .toMatchInlineSnapshot(`
        [
          "HTMLElementAst => HTMLElementAst,template,[object Object],[object Object],[object Object],[object Object]",
          "HTMLTextAst => HTMLTextAst,
          ",
          "HTMLElementAst => HTMLElementAst,div,[object Object],[object Object],[object Object],[object Object]",
          "HTMLTextAst => HTMLTextAst,
        ",
          "HTMLTextAst => HTMLTextAst,
            ",
          "HTMLElementAst => HTMLElementAst,div,[object Object],[object Object]",
          "HTMLTextAst => HTMLTextAst,
          ",
          "HTMLElementAst => HTMLElementAst,input,[object Object],",
        ]
      `);
  });

  test("case2", ({ meta: { name } }) => {
    const sfc = getSFC(name);
    expect(sfc).toMatchInlineSnapshot(`
      {
        "importSnippets": [
          "import { ref } from \\"@vue/reactivity\\"",
        ],
        "script": "const count = ref(0);
      count.value++;
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      console.log(count.value);
      ",
        "setup": true,
        "style": "",
        "template": HTMLElementAst {
          "children": [
            HTMLTextAst {
              "text": "
        ",
              "type": "HTMLTextAst",
            },
            HTMLElementAst {
              "children": [
                HTMLTextAst {
                  "text": "
          ",
                  "type": "HTMLTextAst",
                },
                HTMLElementAst {
                  "children": [
                    HTMLElementAst {
                      "children": [],
                      "props": {
                        ":value": "count",
                      },
                      "tag": "input",
                      "type": "HTMLElementAst",
                    },
                  ],
                  "props": {},
                  "tag": "div",
                  "type": "HTMLElementAst",
                },
                HTMLTextAst {
                  "text": "
        ",
                  "type": "HTMLTextAst",
                },
              ],
              "props": {},
              "tag": "div",
              "type": "HTMLElementAst",
            },
            HTMLTextAst {
              "text": "
      ",
              "type": "HTMLTextAst",
            },
          ],
          "props": {},
          "tag": "template",
          "type": "HTMLElementAst",
        },
        "vapor": true,
      }
    `);
    const code = codegen(sfc);
    writeCode(name, code);
    expect(code).toMatchInlineSnapshot(`
      "import { unref,effect } from \\"@vue/reactivity\\";
      import { ref } from \\"@vue/reactivity\\"

      export default function (props,context){
          const count = ref(0);
        count.value++;
        console.log(count.value);
        console.log(count.value);
        console.log(count.value);
        console.log(count.value);
        console.log(count.value);
        console.log(count.value);
        

          
          var node = {$0:context.parentEl,$1: document.createTextNode(\\"\\\\n  \\"),$2: document.createElement(\\"div\\"),$3: document.createTextNode(\\"\\\\n\\"),$4: document.createTextNode(\\"\\\\n    \\"),$5: document.createElement(\\"div\\"),$6: document.createTextNode(\\"\\\\n  \\"),$7: document.createElement(\\"input\\")};
          node.$0.append(node.$1,node.$2,node.$3);
        
        node.$2.append(node.$4,node.$5,node.$6);
        
        
        node.$5.append(node.$7);
          effect(()=>{effect(()=>{node.$7.setAttribute(\\":value\\",unref(count));});});
      }"
    `);
  });
});
