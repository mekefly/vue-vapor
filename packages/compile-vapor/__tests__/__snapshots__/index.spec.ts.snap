// Vitest Snapshot v1

exports[`index > <template><div><div><div><div><div></div></div></div></div></div></template> 2`] = `
[
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [
              HTMLElementAst {
                "children": [
                  HTMLElementAst {
                    "children": [
                      HTMLElementAst {
                        "children": [],
                        "props": {},
                        "tag": "div",
                        "type": "HTMLElementAst",
                      },
                    ],
                    "props": {},
                    "tag": "div",
                    "type": "HTMLElementAst",
                  },
                ],
                "props": {},
                "tag": "div",
                "type": "HTMLElementAst",
              },
            ],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "template",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [
              HTMLElementAst {
                "children": [
                  HTMLElementAst {
                    "children": [],
                    "props": {},
                    "tag": "div",
                    "type": "HTMLElementAst",
                  },
                ],
                "props": {},
                "tag": "div",
                "type": "HTMLElementAst",
              },
            ],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [
              HTMLElementAst {
                "children": [],
                "props": {},
                "tag": "div",
                "type": "HTMLElementAst",
              },
            ],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
]
`;

exports[`index > <template><div><div><div><div><div></div></div></div></div></div></template> 4`] = `
[
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [
              HTMLElementAst {
                "children": [
                  HTMLElementAst {
                    "children": [
                      HTMLElementAst {
                        "children": [],
                        "props": {},
                        "tag": "div",
                        "type": "HTMLElementAst",
                      },
                    ],
                    "props": {},
                    "tag": "div",
                    "type": "HTMLElementAst",
                  },
                ],
                "props": {},
                "tag": "div",
                "type": "HTMLElementAst",
              },
            ],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "template",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [
              HTMLElementAst {
                "children": [
                  HTMLElementAst {
                    "children": [],
                    "props": {},
                    "tag": "div",
                    "type": "HTMLElementAst",
                  },
                ],
                "props": {},
                "tag": "div",
                "type": "HTMLElementAst",
              },
            ],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [
              HTMLElementAst {
                "children": [],
                "props": {},
                "tag": "div",
                "type": "HTMLElementAst",
              },
            ],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [
          HTMLElementAst {
            "children": [],
            "props": {},
            "tag": "div",
            "type": "HTMLElementAst",
          },
        ],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [
      HTMLElementAst {
        "children": [],
        "props": {},
        "tag": "div",
        "type": "HTMLElementAst",
      },
    ],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
  HTMLElementAst {
    "children": [],
    "props": {},
    "tag": "div",
    "type": "HTMLElementAst",
  },
]
`;
